File: lib/agents/graph.ts
import { StateGraph, START, END } from "@langchain/langgraph";
import { GraphState, type GraphStateType } from "./state";
import type { InputFile } from "../ocr";
import {
  analyticsNode,
  classifyNode,
  expiryNode,
  extractNode,
  languageNode,
  notifyNode,
  ocrNode,
  recommendNode,
  riskNode,
  routeAfterClassify,
  uploadNode,
  validateNode,
} from "./nodes";

// The 11-agent work-permit validation workflow. Each node updates shared state
// and appends safe trace steps; the route handler streams those to the UI.
export function buildGraph() {
  // Node names must not collide with state channel names (ocr, language, risk,
  // expiry, ...), so every node carries a "_step" suffix.
  const graph = new StateGraph(GraphState)
    .addNode("upload_step", uploadNode)
    .addNode("ocr_step", ocrNode)
    .addNode("language_step", languageNode)
    .addNode("classify_step", classifyNode)
    .addNode("extract_step", extractNode)
    .addNode("validate_step", validateNode)
    .addNode("risk_step", riskNode)
    .addNode("expiry_step", expiryNode)
    .addNode("recommend_step", recommendNode)
    .addNode("notify_step", notifyNode)
    .addNode("analytics_step", analyticsNode)
    .addEdge(START, "upload_step")
    .addEdge("upload_step", "ocr_step")
    .addEdge("ocr_step", "language_step")
    .addEdge("language_step", "classify_step")
    .addConditionalEdges("classify_step", routeAfterClassify, { extract: "extract_step", recommend: "recommend_step" })
    .addEdge("extract_step", "validate_step")
    .addEdge("validate_step", "risk_step")
    .addEdge("risk_step", "expiry_step")
    .addEdge("expiry_step", "recommend_step")
    .addEdge("recommend_step", "notify_step")
    .addEdge("notify_step", "analytics_step")
    .addEdge("analytics_step", END);

  return graph.compile();
}

let _compiled: ReturnType<typeof buildGraph> | null = null;
export function getGraph() {
  if (!_compiled) _compiled = buildGraph();
  return _compiled;
}

export function createInitialState(files: InputFile[], message: string): Partial<GraphStateType> {
  return { files, message };
}

File: backend/app/agents/graph.py
"""The 11-agent work-permit validation workflow (LangGraph StateGraph).

Node names carry a ``_step`` suffix because LangGraph forbids node names that
collide with state channel names (e.g. ``ocr``, ``risk``, ``expiry``).
"""
from __future__ import annotations

from langgraph.graph import END, START, StateGraph

from . import nodes
from .state import GraphState

_compiled = None


def build_graph():
    graph = StateGraph(GraphState)
    graph.add_node("upload_step", nodes.upload_node)
    graph.add_node("ocr_step", nodes.ocr_node)
    graph.add_node("language_step", nodes.language_node)
    graph.add_node("classify_step", nodes.classify_node)
    graph.add_node("extract_step", nodes.extract_node)
    graph.add_node("validate_step", nodes.validate_node)
    graph.add_node("risk_step", nodes.risk_node)
    graph.add_node("expiry_step", nodes.expiry_node)
    graph.add_node("recommend_step", nodes.recommend_node)
    graph.add_node("notify_step", nodes.notify_node)
    graph.add_node("analytics_step", nodes.analytics_node)

    graph.add_edge(START, "upload_step")
    graph.add_edge("upload_step", "ocr_step")
    graph.add_edge("ocr_step", "language_step")
    graph.add_edge("language_step", "classify_step")
    graph.add_conditional_edges(
        "classify_step",
        nodes.route_after_classify,
        {"extract": "extract_step", "recommend": "recommend_step"},
    )
    graph.add_edge("extract_step", "validate_step")
    graph.add_edge("validate_step", "risk_step")
    graph.add_edge("risk_step", "expiry_step")
    graph.add_edge("expiry_step", "recommend_step")
    graph.add_edge("recommend_step", "notify_step")
    graph.add_edge("notify_step", "analytics_step")
    graph.add_edge("analytics_step", END)

    return graph.compile()


def get_graph():
    global _compiled
    if _compiled is None:
        _compiled = build_graph()
    return _compiled
